From 5170df2d3fa9c69e855c4b7c2b084ebfd92d5200 Mon Sep 17 00:00:00 2001
From: Thomas Piart <piartt@gmail.com>
Date: Wed, 21 Jul 2021 23:34:54 +0200
Subject: [PATCH] Escape HTML  by default in list_tag ecape the Tag values. It
 should be text, not HTML This is a breaking change and documentation should
 be updated to explain that when using function transform, value should be
 escaped if needed

---
 lib/plugins/helper/list_tags.js   |  4 ++--
 test/scripts/helpers/list_tags.js | 38 +++++++++++++++++++++++++++++++
 2 files changed, 40 insertions(+), 2 deletions(-)

diff --git a/lib/plugins/helper/list_tags.js b/lib/plugins/helper/list_tags.js
index 252eb62ef..6d5685051 100644
--- a/lib/plugins/helper/list_tags.js
+++ b/lib/plugins/helper/list_tags.js
@@ -1,6 +1,6 @@
 'use strict';
 
-const { url_for } = require('hexo-util');
+const { url_for, escapeHTML } = require('hexo-util');
 
 function listTagsHelper(tags, options) {
   if (!options && (!tags || !Object.prototype.hasOwnProperty.call(tags, 'length'))) {
@@ -59,7 +59,7 @@ function listTagsHelper(tags, options) {
       result += `<li class="${liClass}">`;
 
       result += `<a class="${aClass}" href="${url_for.call(this, tag.path)}${suffix}" rel="tag">`;
-      result += transform ? transform(tag.name) : tag.name;
+      result += transform ? transform(tag.name) : escapeHTML(tag.name);
       result += '</a>';
 
       if (showCount) {
diff --git a/test/scripts/helpers/list_tags.js b/test/scripts/helpers/list_tags.js
index f4d3ed62d..7e07280fb 100644
--- a/test/scripts/helpers/list_tags.js
+++ b/test/scripts/helpers/list_tags.js
@@ -205,3 +205,41 @@ describe('list_tags', () => {
     ].join(''));
   });
 });
+
+describe('list_tags transform', () => {
+  const Hexo = require('../../../lib/hexo');
+  const hexo = new Hexo(__dirname);
+  const Post = hexo.model('Post');
+
+  const ctx = {
+    config: hexo.config
+  };
+
+  const listTags = require('../../../lib/plugins/helper/list_tags').bind(ctx);
+
+  before(async () => {
+    await hexo.init();
+    const posts = await Post.insert([
+      {source: 'foo', slug: 'foo'}
+    ]);
+
+    // TODO: Warehouse needs to add a mutex lock when writing data to avoid data sync problem
+    await Promise.all([
+      ['bad<b>HTML</b>']
+    ].map((tags, i) => posts[i].setTags(tags)));
+
+    hexo.locals.invalidate();
+    ctx.site = hexo.locals.toObject();
+  });
+
+  // no transform should escape HTML
+  it('no transform', () => {
+    const result = listTags();
+
+    result.should.eql([
+      '<ul class="tag-list" itemprop="keywords">',
+      '<li class="tag-list-item"><a class="tag-list-link" href="/tags/bad-b-HTML-b/" rel="tag">bad&lt;b&gt;HTML&lt;&#x2F;b&gt;</a><span class="tag-list-count">1</span></li>',
+      '</ul>'
+    ].join(''));
+  });
+});
